varinfo/generate_umm_var.py

```diff
@@ -0,0 +1,92 @@
+
+from cmr import CMR_UAT
+
+from varinfo import VarInfoFromNetCDF4
+from varinfo.cmr_search import (CmrEnvType, download_granule, get_granule_link,
+                                get_granules)
+from varinfo.umm_var import get_all_umm_var, publish_all_umm_var
+
+
+# Custom return type: either a list of UMM-Var JSON (a list of dictionaries),
+# or a list of strings (either concept IDs or error strings).
+UmmVarReturnType = List[Union[Dict, str]]
+
+
+def generate_collection_umm_var(collection_concept_id: str,
+                                auth_header: str,
+                                cmr_env: CmrEnvType = CMR_UAT,
+                                publish: bool = False) -> UmmVarReturnType:
+    """ Run all the of the functions for downloading and publishing
+        a UMM-Var entry to CMR given:
+
+        * collection_concept_id: Concept ID for collection that variables have
+          been generated for.
+        * cmr_env: URLs for CMR environments (OPS, UAT, and SIT)
+        * auth_header: Authorization HTTP header, containing either:
+          - A LaunchPad token: 'Authorization: <token>'
+          - An EDL bearer token: 'Authorization: Bearer <token>'
+        * publish: Optional argument determining whether to publish the
+          generated UMM-Var records to the indicated CMR instance. Defaults to
+          False.
+
+        Note - if attempting to publish to CMR, a LaunchPad token must be used.
+
+    """
+    granule_response = get_granules(collection_concept_id, cmr_env=cmr_env,
+                                    auth_header=auth_header)
+
+    # Get the data download URL for the most recent granule (NetCDF-4 file)
+    granule_link = get_granule_link(granule_response)
+
+    with TemporaryDirectory() as temp_dir:
+        # Download file to runtime environment
+        local_granule = download_granule(granule_link,
+                                         auth_header,
+                                         out_directory=temp_dir)
+
+        # Parse the granule with VarInfo to map all variables and relations:
+        var_info = VarInfoFromNetCDF4(local_granule)
+
+        # Generate all the UMM-Var records:
+        all_umm_var_records = get_all_umm_var(var_info)
+
+    if publish:
+        # Publish to CMR and construct an output object that is a list of
+        # strings. These strings will be either variable concept IDs or
+        # error messages returned from CMR.
+        publication_response = publish_all_umm_var(collection_concept_id,
+                                                   all_umm_var_records,
+                                                   auth_header, cmr_env)
+
+        return_value = [variable_response
+                        if is_variable_concept_id(variable_response)
+                        else ': '.join([variable_name, variable_response])
+                        for variable_name, variable_response
+                        in publication_response.items()]
+    else:
```

Comment:
Why not just return `all_umm_var_records` I don't see what the point of making it into a list is.

Reply:
I think we agreed with William and Ryan that they are expecting a bunch of concept IDs, and that we could add a verbose mode later on that would include both.

This list comprehension is doing a little bit more, though. If the message gotten back from `publish_umm_var` for the variable is indeed a variable concept ID, then that's what is returned as the list element. If there was an error, then the error message is combined with the variable long name to give the user a clearer indication of what failed. So, it'd look something like:

```
return_value = [
    'V1234567890-EEDTEST',
    '/full/variable/path: This is the error message from CMR'
]
```

Reply:
(Mentioned in other thread - will add clarifying comment to make this easier to grok in the code)